src/autopas/LogicHandler.h

```diff
@@ -1066,17 +1218,178 @@ void LogicHandler<Particle>::remainderHelperBufferHaloBuffer(
+              const std::lock_guard<std::mutex> lock(*_spacialLocks[lockCoords[0]][lockCoords[1]][lockCoords[2]]);
+              f->AoSFunctor(p3, p1, p2, false);
+            }
+          },
+          min, max, IteratorBehavior::ownedOrHalo);
+    }
+  }
+
+#if SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_TRACE
+  timerBufferBufferContainer.stop();
+  timerBufferContainerContainer.start();
+#endif
+
+  // Step 3: 3-body interactions of 1 buffer particle and 2 container particles
+  // todo: parallelize without race conditions
+  // AUTOPAS_OPENMP(parallel for shared(bufferParticles))
+  for (auto i = 0; i < numTotal; ++i) {
+    Particle &p1 = *bufferParticles[i];
+    const auto pos = p1.getR();
+    const auto boxmin = pos - cutoff;
+    const auto boxmax = pos + cutoff;
+
+    auto p2Iter = container.getRegionIterator(
+        boxmin, boxmax, IteratorBehavior::ownedOrHalo | IteratorBehavior::forceSequential, nullptr);
+    for (; p2Iter.isValid(); ++p2Iter) {
+      Particle &p2 = *p2Iter;
+
+      auto p3Iter = p2Iter;
+      ++p3Iter;
+      for (; p3Iter.isValid(); ++p3Iter) {
+        Particle &p3 = *p3Iter;
+
+        if constexpr (newton3) {
+          f->AoSFunctor(p1, p2, p3, true);
+        } else {
+          if (i < numOwnedBufferParticles) {
+            f->AoSFunctor(p1, p2, p3, false);
+          }
+          // no need to calculate force enacted on a halo
+          if (p2.isOwned()) {
+            f->AoSFunctor(p2, p1, p3, false);
+          }
+          if (p3.isOwned()) {
+            f->AoSFunctor(p3, p1, p2, false);
+          }
+        }
+      }
+    }
+  }
+
+#if SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_TRACE
+  timerBufferContainerContainer.stop();
+#endif
+
+  AutoPasLog(TRACE, "Timer Buffer <-> Buffer <-> Buffer       : {}", timerBufferBufferBuffer.getTotalTime());
+  AutoPasLog(TRACE, "Timer Buffer <-> Buffer <-> Container    : {}", timerBufferBufferContainer.getTotalTime());
+  AutoPasLog(TRACE, "Timer Buffer <-> Container <-> Container : {}", timerBufferContainerContainer.getTotalTime());
+}
+
 template <typename Particle>
 template <class Functor>
 std::tuple<Configuration, std::unique_ptr<TraversalInterface>, bool> LogicHandler<Particle>::selectConfiguration(
-    Functor &functor) {
+    Functor &functor, InteractionTypeOption::Value interactionType) {
   bool stillTuning = false;
   Configuration configuration{};
   std::optional<std::unique_ptr<TraversalInterface>> traversalPtrOpt{};
+  auto &autoTuner = _autoTunerRefs[interactionType];
+
   // if this iteration is not relevant take the same algorithm config as before.
   if (not functor.isRelevantForTuning()) {
     stillTuning = false;
-    configuration = _autoTuner.getCurrentConfig();
+    configuration = autoTuner->getCurrentConfig();
+    // This is needed to synchronize container types between multiple interaction types.
+    if (_containerSelector.getCurrentContainer().getContainerType() != configuration.container) {
+      _containerSelector.selectContainer(
```

Comment:
Please expand this comment. I don't fully get it but invoking `selectContainer()` is potentially expensive so it should be properly explained why this is necessary and what case you are actually handling here.

Reply:
I changed the comment and added an issue (#871)